### PR TITLE
[AIRFLOW-1438] Change batch size per query

### PR DESCRIPTION
**Pull request by saguziel**
Dear Airflow maintainers,

Please accept this PR. I understand that it will not be reviewed until I have checked off all the steps below!

### JIRA
- [x] My PR addresses the following [Airflow JIRA](https://issues.apache.org/jira/browse/AIRFLOW/) issues and references them in the PR title. For example, "[AIRFLOW-XXX] My Airflow PR"
    - https://issues.apache.org/jira/browse/AIRFLOW-1438

### Description
- [x] Here are some details about my PR, including screenshots of any UI changes: Adds a tunable batch limit to scheduler queries

### Tests
- [x] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason: Will add some tests

### Commits
- [x] My commits all reference JIRA issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
    1. Subject is separated from body by a blank line
    2. Subject is limited to 50 characters
    3. Subject does not end with a period
    4. Subject uses the imperative mood ("add", not "adding")
    5. Body wraps at 72 characters
    6. Body explains "what" and "why", not "how"

### CODE REVIEW

**Comment by saguziel** [comment]
to: @bolkedebruin @aoen @edgarRd

Code is done. Will do one last pass to verify the SQL is what I want

File: airflow/config_templates/default_airflow.cfg [hunk 1]
```diff
@@ -325,6 +325,11 @@ scheduler_zombie_task_threshold = 300
 # DAG definition (catchup)
 catchup_by_default = True
 
+# This changes the batch size of queries in the scheduling main loop.
+# This depends on query length limits and how long you are willing to hold locks.
+# 0 for no limit
+max_tis_per_query = 256
```
Comment (bolkedebruin) [inline airflow/config_templates/default_airflow.cfg]: Is there is reason to set this to 256? It changes the default behaviour?

**Comment by bolkedebruin** [comment]
Reading through the code around it I'm wondering why we are holding locks for long (where are we doing so? couldn't really find it) and that the code in this section seems to get really complex and I am not sure why it is.

Apart from the default setting, LGTM

File: airflow/config_templates/default_airflow.cfg (hunk 1 above)
Comment (saguziel) [reply airflow/config_templates/default_airflow.cfg]: Fair enough

**Comment by saguziel** [comment]
Things are queried in batch. Locks are held not for super long but maybe the order of a few seconds. The complexity is for performance

File: airflow/config_templates/default_airflow.cfg [hunk 2]
```diff
@@ -325,6 +325,11 @@ scheduler_zombie_task_threshold = 300
 # DAG definition (catchup)
 catchup_by_default = True
 
+# This changes the batch size of queries in the scheduling main loop.
+# This depends on query length limits and how long you are willing to hold locks.
+# 0 for no limit
+max_tis_per_query = 0
```
Comment (aoen) [inline airflow/config_templates/default_airflow.cfg:331]: what about max_tis_per_loop or max_tis_examined_per_loop? I feel like the current name isn't self-explanatory enough.

File: airflow/jobs.py [hunk 3]
```diff
@@ -1289,15 +1299,32 @@ def _execute_task_instances(self,
...
-        tis_with_state_changed = self._change_state_for_executable_task_instances(
-            executable_tis,
-            states,
-            session=session)
-        self._enqueue_task_instances_with_queued_state(
-            simple_dag_bag,
-            tis_with_state_changed)
-        session.commit()
-        return len(tis_with_state_changed)
+        if self.max_tis_per_query == 0:
```
Comment (aoen) [inline airflow/jobs.py:1302]: Can we deduplicate this logic? i.e. tis_per_query == 0 means chunks just has one element (then reuse the rest of the logic).

File: tests/jobs.py [hunk 4]
```diff
@@ -1037,6 +1037,44 @@ def test_execute_task_instances(self):
...
+        task_id_1 = 'dummy_task'
+        task_id_2 = 'dummy_task_2'
+        # important that len(tasks) is less than concurrency
+        # because before scheduler._execute_task_instances would only
+        # check the num tasks once so if concurrency was 3,
+        # we could execute arbitrarily many tasks in the second run
+        dag = DAG(dag_id=dag_id, start_date=DEFAULT_DATE, concurrency=16)
+        task1 = DummyOperator(dag=dag, task_id=task_id_1)
+        task2 = DummyOperator(dag=dag, task_id=task_id_2)
+        dagbag = SimpleDagBag([dag])
```
Comment (aoen) [inline tests/jobs.py:1051]: Should test the codepath where batch size is 0. Even if other tests test it tests should be explicit (e.g. other tests might not be testing the edge cases we want tested).

File: airflow/jobs.py [hunk 5]
```diff
@@ -1146,6 +1147,15 @@ def _find_executable_task_instances(self, simple_dag_bag, states, session=None):
             ["{}".format(x) for x in executable_tis])
         self.logger.info("Setting the follow tasks to queued state:\n\t{}"
                          .format(task_instance_str))
+        # so these dont expire on commit
```
Comment (aoen) [inline airflow/jobs.py:1150]: I think this comment could be made a bit clearer

File: airflow/config_templates/default_airflow.cfg (hunk 2 above)
Comment (saguziel) [reply airflow/config_templates/default_airflow.cfg:331]: That name is worse IMO but it sounds like it is per scheduler loop, which it is not.

File: tests/jobs.py (hunk 4 above)
Comment (saguziel) [reply tests/jobs.py:1051]: yeah fair enoguh defaults could change

File: airflow/config_templates/default_airflow.cfg (hunk 2 above)
Comment (aoen) [reply airflow/config_templates/default_airflow.cfg:331]: maybe add an adjective to query? max_tis_per_main_query? This sounds like it's a global to all ti queries in the scheduler.

Comment (saguziel) [reply airflow/config_templates/default_airflow.cfg:331]: I'm not sure. I think no matter what you change the name too it is kind of unclear, and the comment explains it anyway.

Comment (aoen) [reply airflow/config_templates/default_airflow.cfg:331]: I think I'd rather have something more verbose here if it makes it clearer. I think expecting users to find the default airflow config to get an idea of what this does isn't great. I guess it's a bit of a nit, but anything that's exposed to users should have a bit of care put in to it.

**Pull request closed without merging**